File: tree/path_and_sum/all_path.go
package pathandsum

import (
	"fmt"
	"trouble_solver/tree"
)

func binaryTreePaths(root *tree.TreeNode) []string {
	if root == nil {
		return []string{}
	}

	var ans []string
	var findPath func(root *tree.TreeNode, path string)
	findPath = func(root *tree.TreeNode, path string) {
		if root == nil {
			return
		}

		str := fmt.Sprintf("%d", root.Val)
		path += str
		if root.Left == nil && root.Right == nil {
			ans = append(ans, path)
			return
		}

		findPath(root.Left, path+"->")
		findPath(root.Right, path+"->")
	}
	return ans
}

// 很明显，能够看出这是前序遍历，所以也能够用栈来迭代实现
func binaryTreePaths2(root *tree.TreeNode) []string {
	if root == nil {
		return []string{}
	}
	stackForNode, stackForPath := []*tree.TreeNode{}, []string{}
	var ans []string
	stackForNode = append(stackForNode, root)
	stackForPath = append(stackForPath, fmt.Sprintf("%d", root.Val))

	for len(stackForNode) > 0 {
		length, lenForStackPath := len(stackForNode), len(stackForPath)
		node := stackForNode[length-1]
		stackForNode = stackForNode[:length-1]

		if node.Left == nil && node.Right == nil {
			ans = append(ans, stackForPath[lenForStackPath-1])
			stackForPath = stackForPath[:lenForStackPath-1]
			continue
		}

		path := stackForPath[lenForStackPath-1]
		stackForPath = stackForPath[:lenForStackPath-1]
		if node.Left != nil {
			stackForNode = append(stackForNode, node.Left)
			stackForPath = append(stackForPath, path+"->"+fmt.Sprintf("%d", node.Left.Val))

		}
		if node.Right != nil {
			stackForNode = append(stackForNode, node.Right)
			stackForPath = append(stackForPath, path+"->"+fmt.Sprintf("%d", node.Right.Val))
		}
	}

	return ans

}

File: tree/path_and_sum/AllPath.java
/*
 * @Author: yangxcc
 * @version: 1.0
 * @Date: 2022-10-27 17:35:57
 * @LastEditTime: 2023-03-03 15:58:14
 */
package tree.path_and_sum;

import java.util.ArrayList;
import java.util.List;

import tree.TreeNode;

/**
 * leetcode 257 simple 二叉树的所有路径
 * 
 * 给你一个二叉树的根节点 root ，按 任意顺序 ，返回所有从根节点到叶子节点的路径。
 * 叶子节点 是指没有子节点的节点。
 */
public class AllPath {
    List<List<Integer>> ans = new ArrayList<>();

    public List<List<Integer>> findAllPath(TreeNode root) {
        if (root == null) {
            return new ArrayList<>();
        }   

        helper(root, new ArrayList<>());

        return ans;
    }

    public void helper(TreeNode root, List<Integer> path) {
        if (root == null) {
            return;
        }

        path.add(root.val);
        
        if (root.left == null && root.right == null) {
            ans.add(new ArrayList<>(path));
            return;
        }

        helper(root.left, path);
        helper(root.right, path);
    }
}